### PR TITLE
V2 empty seq

### PR DESCRIPTION
**Pull request by CanBerker**
**Second take** on the empty sequence problem:

First two commits fix the out of bounds and illegal next calling issues.

Last commit is experimental and fixes the nullpointer exceptions occurring in the spark framework. If result.iterator() contains null elements which means that the result of the query is an empty sequence, an exception is thrown. By removing the said null elements, no exceptions are met. However, I'm worried that this solution might be crude and there might be corner cases that I'm missing.

Following queries which provided the respective exceptions now work as intended and return empty sequences. Related tests are also added.

let $o := (()) return $o()
✔ [ERROR] Job aborted due to stage failure: ..... java.lang.IndexOutOfBoundsException:

let $o := (()) return size($o)
✔ [ERROR] Job aborted due to stage failure: ..... java.lang.IndexOutOfBoundsException:

let $o := (1, 2, "a", [], 5e+3) return descendant-objects($o)
✔ [ERROR] Job aborted due to stage failure: ..... java.lang.NullPointerException

let $o := ([]) return flatten($o)
✔ [ERROR] Job aborted due to stage failure: ..... java.lang.NullPointerException

let $o := ({}) return descendant-arrays($o)
✔ [ERROR] Job aborted due to stage failure: ..... java.lang.NullPointerException

### CODE REVIEW

File: src/main/java/sparksoniq/jsoniq/runtime/iterator/RuntimeIterator.java [hunk 1]
```diff
@@ -132,11 +132,15 @@ public IteratorMetadata getMetadata() {
...
+        Item result = null;
+        if(iterator.hasNext()) {
+            result = iterator.next();
+            if (iterator.hasNext()) {
+                throw nonAtomicException;
+            }
+        }
         iterator.close();
-        if (!(type.isInstance(result)))
+        if (result != null && !(type.isInstance(result)))
```
Comment (ghislainfourny) [inline src/main/java/sparksoniq/jsoniq/runtime/iterator/RuntimeIterator.java]: Yes, I think that this is fine this way. This function is purely an internal helper that isn't exposed to the outside world, so you can use the Java null as a special return value that says "there is no such single item".

File: src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/arrays/ArraySizeFunctionIterator.java [hunk 2]
```diff
@@ -43,6 +43,9 @@ public Item next() {
             this._hasNext = false;
             RuntimeIterator arrayIterator = this._children.get(0);
```
Comment (ghislainfourny) [inline src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/arrays/ArraySizeFunctionIterator.java]: This is another example of when you can/should move the array==null test as well as the two lines above to the end of an overriden open() method (after the super call). That way, if array is tested to be the Java null, you can set this._hasNext to false already when opening the iterator and next() won't even be called by the consumer.

In many iterators, a similar thing needs to be done, i.e., move some code from next() to an overriden open() in order to know right away whether an empty sequence is returned or not. This approach is very standard in pull-based querying engines (Zorba does it as well).

File: src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/object/ObjectDescendantFunctionIterator.java [hunk 3]
```diff
@@ -20,7 +20,6 @@ public ObjectDescendantFunctionIterator(List<RuntimeIterator> arguments, Iterato
     public Item next() {
         if (this._hasNext) {
             if (results == null) {
```
Comment (ghislainfourny) [inline src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/object/ObjectDescendantFunctionIterator.java]: Likewise -- this initialization should be done at the end of an overriden open() method, as well as setting _hasNext to false if there are no results at all.

File: src/main/java/sparksoniq/jsoniq/runtime/iterator/primary/VariableReferenceIterator.java [hunk 4]
```diff
@@ -64,9 +61,13 @@ public void open(DynamicContext context){
             throw new IteratorFlowException("Variable reference iterator already open", getMetadata());
         this._currentDynamicContext = context;
         this.currentIndex = 0;
-        this.items = null;
-        this._hasNext = true;
-
+        this.items = this._currentDynamicContext.getVariableValue(this._variableName);
+        if(items.size() == 0) {
+            this._hasNext = false;
```
Comment (ghislainfourny) [inline src/main/java/sparksoniq/jsoniq/runtime/iterator/primary/VariableReferenceIterator.java]: Yes, exactly like this :-) This is the general idea.

File: src/main/java/sparksoniq/spark/closures/ReturnFlatMapClosure.java [hunk 5]
```diff
@@ -43,6 +44,7 @@ public ReturnFlatMapClosure(RuntimeIterator expression) {
         while (_expression.hasNext())
             result.add(_expression.next());
         _expression.close();
+        result.removeIf(Objects::isNull);
```
Comment (ghislainfourny) [inline src/main/java/sparksoniq/spark/closures/ReturnFlatMapClosure.java]: I understand the temptation to do this, but this is risky -- if there are nulls (which there should never be), you will no longer know: it is a bit like putting the dust under the carpet or switching off the fire alarm rather than fighting the fire :-) It is better practice to let the exception be thrown so that you can fix all the places were a Java null lands in here (that way, you will catch all iterators in which a move of code from next() to open() is necessary).

File: src/main/java/sparksoniq/spark/iterator/SparkRuntimeIterator.java [hunk 6]
```diff
@@ -70,6 +70,10 @@ public Item next(){
                 result = _rdd.collect();
         }
 
+        if(result.size() == 0){
```
Comment (ghislainfourny) [inline src/main/java/sparksoniq/spark/iterator/SparkRuntimeIterator.java]: Likewise, this and the above initialization should go to open().

**Comment by CanBerker** [comment]
Calculations in iterators have been moved to the overrides of open() function. Since I've worked with each iterator in a different commit, there is a huge number of them. I can squash them together into a single commit after receiving your approval to keep a cleaner commit history.

File: src/main/java/sparksoniq/jsoniq/runtime/iterator/CommaExpressionIterator.java [hunk 7]
```diff
@@ -17,59 +17,57 @@
...
-        currentChild = null;
-    }
+        this._currentIndex = 0;
+        this.results = new ArrayList<>();
 
-    @Override
-    public void reset(DynamicContext context){
-        super.reset(context);
-        currentChild = null;
+        // this._children.forEach(c -> c.reset(this._currentDynamicContext));
```
Comment (wscsprint3r) [inline src/main/java/sparksoniq/jsoniq/runtime/iterator/CommaExpressionIterator.java]: I understand the problem you are trying to solve but I think this is taking it a bit too far. Comma expressions, arrays and any Spark Iterator might be dealing with a very large amount of results and pre-computing all of the results in the open() method will potentially remove all the laziness resulting in significantly lower performance. @ghislainfourny what do you think?

File: src/main/java/sparksoniq/spark/iterator/SparkRuntimeIterator.java (hunk 6 above)
Comment (wscsprint3r) [reply src/main/java/sparksoniq/spark/iterator/SparkRuntimeIterator.java]: @ghislainfourny  that will cause the RDD computation to be executed before next() is actually called. Do we want this?

File: src/main/java/sparksoniq/jsoniq/runtime/iterator/CommaExpressionIterator.java (hunk 7 above)
Comment (wscsprint3r) [reply src/main/java/sparksoniq/jsoniq/runtime/iterator/CommaExpressionIterator.java]: Ideally we would only peek at the first item in the open() method in order to return false for hasNext, instead of computing all the results, at least in some cases.

**Comment by ghislainfourny** [comment]
As discussed today: this PR will need to be split into many, smaller PRs (typically one per iterator, possibly grouped when semantically similar), and we will have a discussion on each one of them separately and in parallel.

File: src/main/java/sparksoniq/jsoniq/runtime/iterator/CommaExpressionIterator.java (hunk 7 above)
Comment (ghislainfourny) [reply src/main/java/sparksoniq/jsoniq/runtime/iterator/CommaExpressionIterator.java]: Stefan has a good point -- we discussed this yesterday with Can.

The idea is to be as lazy as possible and *only* do the computations in open that are absolutely necessary (such as getting a lookahead of one item for hasNext() to return the correct result).

In some (rare) cases, it is necessary to compute everything in open. This is known in literature, and called "materialization". For example, this is the case with the reverse function: you can only return the first item by consuming the entire input sequence. This should be done only if absolutely necessary because of the semantics of the iterator.

This will partially be the case with local group by/order by execution, as the first tuple can only be returned after consuming the entire input stream.

Comment (ghislainfourny) [reply src/main/java/sparksoniq/jsoniq/runtime/iterator/CommaExpressionIterator.java]: Regarding iterators with an underlying RDD: the design pretty much depends on the Spark API. If the Spark API allows us to "peek" at the RDD to see if it is empty, then it is not necessary to precompute in open() before haxNext() is called. If the Spark API allows getting the items one by one, this is also not necessary. If all Spark allows is "taking" all items at once, then materialization is necessary (but of course, in practice, this will only be done for small sequences, and an error can be thrown for bigger sequences).
